packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,246 @@
+                                        &outdegrees, &weighted );
+        assert( weighted == 0 );
+        assert( outdegrees == (int)destinations.size() );
+
+        std::vector<int> sources( indegrees );
+        MPI_Dist_graph_neighbors( _comm_dist_graph, indegrees,
+                                  notNullPtr( sources.data() ), MPI_UNWEIGHTED,
+                                  outdegrees, notNullPtr( destinations.data() ),
+                                  MPI_UNWEIGHTED );
+
+        _src_counts.resize( indegrees );
+        MPI_Neighbor_alltoall( _dest_counts.data(), 1, MPI_INT,
+                               _src_counts.data(), 1, MPI_INT,
+                               _comm_dist_graph );
+
+        _src_offsets.resize( indegrees + 1 );
+        // exclusive scan
+        _src_offsets[0] = 0;
+        for ( int i = 0; i < indegrees; ++i )
+            _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+
+#if defined( REORDER_RECV )
+        int comm_size = -1;
+        MPI_Comm_size( _comm, &comm_size );
+        _permute_recv.resize( _src_offsets.back() );
+        int offset = 0;
+        for ( int i = 0; i < comm_size; ++i )
+        {
+            auto const it = std::find( sources.begin(), sources.end(), i );
+            if ( it != sources.end() )
+            {
+                int const j = std::distance( sources.begin(), it );
+                std::iota( &_permute_recv[offset],
+                           &_permute_recv[offset] + _src_counts[j],
+                           _src_offsets[j] );
+                offset += _src_counts[j];
+            }
+        }
+        assert( offset == (int)_permute_recv.size() );
+#endif
+
+        return _src_offsets.back();
+    }
+    template <typename View>
+    void doPostsAndWaits( typename View::const_type const &exports,
+                          size_t numPackets, View const &imports ) const
+    {
+        using ValueType = typename View::value_type;
+        static_assert( View::rank == 1, "" );
+
+        std::vector<int> dest_counts = _dest_counts;
+        std::vector<int> dest_offsets = _dest_offsets;
+        std::vector<int> src_counts = _src_counts;
+        std::vector<int> src_offsets = _src_offsets;
+        for ( auto pv :
+              {&dest_counts, &dest_offsets, &src_counts, &src_offsets} )
+            for ( auto &x : *pv )
+                x *= numPackets * sizeof( ValueType );
+
+        std::vector<ValueType> dest_buffer( exports.size() );
+        std::vector<ValueType> src_buffer( imports.size() );
+
+        assert( (size_t)src_offsets.back() ==
+                imports.size() * sizeof( ValueType ) );
+        assert( (size_t)dest_offsets.back() ==
+                exports.size() * sizeof( ValueType ) );
+
+        for ( int i = 0; i < _dest_offsets.back(); ++i )
+            std::copy( &exports[numPackets * i],
+                       &exports[numPackets * i] + numPackets,
+                       &dest_buffer[numPackets * _permute[i]] );
```

Comment:
How does this work when `exports` in on the device?

Reply:
(looks like my reply has been lost...)
I want to add a static assertion that the memory space is accessible from the host.  They are still a few rough edges with `sendAcrossNetwork()` and the distributor construction that need to be reworked (e.g. unnecessary copies)

Reply:
I added a static assertion that the underlying data in the view is accessible from the host.